EP: IPSec Central CA (#102)
submariner-io/enhancements: submariner/ipsec-cert-issuer.md

# IPSec Certificate Issuer

## Summary

> Submariners current security arrangements, namely using an identical pre-shared key in

Reviewer: ```suggestion
Submariner's current security arrangements, namely using an identical pre-shared key in
```

> all deployments will fail security assessment in most environments which have to comply

Reviewer: ```suggestion
all deployments, will fail security assessment in most environments which have to comply
```

> all deployments will fail security assessment in most environments which have to comply

Reviewer: ```suggestion
all clusters in a ClusterSet will fail security assessment in most environments which have to comply
```

Is this equivalent to what you mean? If so, it might be more clear to be a bit more explicit.

with a pre-existing security policy and rules on using VPNs in general and IPSEC in particular.

> In order to comply the minimum enhancement that needs to be implemented is the use

Reviewer 2: Can we also discuss alternatives?

I see CA is not always preferred to PSK [1] so I think we can also consider the alternative security model and pros and cons of each.

[1] https://community.cisco.com/t5/vpn/ipsec-authentication-using-x509-certificates/m-p/3215190/highlight/true#M119210

Reviewer 3: The alternative is to ensure we use one PSK per connected cluster. This would require inventing our own ways of authenticating clients and revoking PSKs; CAs provide standardised ways of handling all this.

Reviewer 2: I think it would be useful to capture this in the EP

> In order to comply the minimum enhancement that needs to be implemented is the use

Reviewer: ```suggestion
In order to comply, the minimum enhancement that needs to be implemented is the use
```
IMHO

> In order to comply the following minimum enhancement that needs to be implemented is the use

Author: ```suggestion
In order to comply the minimum enhancement that needs to be implemented is the use
```

> of X509 certificates per Gateway. This will require the introduction of a Certificate Authority

Reviewer 4: I most likely missed it, but am unclear on how would you establish trust chain with the CA? Its certificate bundle should be present on each GW so it needs to be sync'ed.

Author: The CAs cert can be synced to the clusters in the same way the self signed broker cert is synced today (via a volume mounted secret). On GW startup this cert could be read from the volume and then loaded into /etc/pki/tls/certs/ca-bundle.crt
I will add more details on this

> of X509 certificates per Gateway. This will require the introduction of a Certificate Authority

Reviewer: ```suggestion
of X.509 certificates per Gateway. This will require the introduction of a Certificate Authority
```

Seems to be the consistent and [official](https://datatracker.ietf.org/doc/html/rfc5280) usage.

(CA) that's capable of signing Certificate Signing Requests (CSRs) for Gateways in the
`ClusterSet`. This CA will be serviced by a Syncer that will act as a proxy between the Certificate
Requester and the CA itself.

## Proposal

> This enhancement proposes to use a `ClusterSet` wide `Certificate Authority` and a `Syncer`

Reviewer: ```suggestion
This enhancement proposes to use a `ClusterSet`-wide `Certificate Authority` and a `Syncer`
```

> that are co-located with the Broker to service Gateway. At a high level this requires:

Reviewer 2: Does it have to be co-located with the broker?

> that is co-located with the Broker to service Gateway. At a high level this requires:

Author: ```suggestion
that are co-located with the Broker to service Gateway. At a high level this requires:
```

* A Certificate Authority that can be used across the ClusterSet. Ideally this authority resides
on the Broker node.
* A CA_CSR_Syncer on the Broker Node.
* A new CRD to encapsulate a Submariner CSR (towards this centralized CA), and the result (towards

> the Gateway). Gateways will fill the CRD with relevant information (a base64 encoded string of a PEM

Reviewer: ```suggestion
the Gateway). Gateways will fill the CRD with relevant information (a base64-encoded string of a PEM-
```

encoded certificate request) and publish the CR to the Kube API.
* A GW_CSR_Syncer on the gateway Node.

The general idea is that, when a Broker is installed, a CA and CA_CSR_Syncer are also installed on
the Broker node. Then, when a Submariner gateway starts up it will launch a GW_CSR_Syncer, it will

> also create a CSR, base64 encode it and send it via a Submariner CSR CRD towards the Broker node CA.

Reviewer 3: Could we use general-purpose Kubernetes CSRs instead? _I.e._ have the client gateway directly submit a CSR to the broker cluster, instead of going through syncers.

Author: Just so we are clear are you suggesting:

1. Using the Broker Node Kubernetes API/CP as the dedicate 
ClusterSet CA and getting rid of the proposed Submariner CSR CRD? so then the syncer on the local/GW node forwards/exports CSRs to the broker node (via federator) and the rest of the process from there on out is the same (sign, publish result, publish back to GW)? OR

2. Just using the Kubernetes CSRs but still using something else for the CA such as cert-manager? 

I guess for both options 1 and 2 I need to look into the following questions:

- How would we distinguish between local (Pod/svc) CSRs vs the GW CSRs to make sure that the GW CSRs are not signed locally... and
- does the K8s CSR facility do any sort of verification of svcs/pods in the cluster before signing a cert?

The reason why I suggested cert-manager was that you can really configure different CAs behind the cert-manager API I also thought it was what ACM uses and potentially for the GW verification we may want to have an independent CA to what's provided in the scope of a local cluster.

Reviewer 4: Using the `spec.signerName` may offer a way to distinguish whom should be signing the CSR.

Reviewer 4: Some CSR verification in https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers.

Reviewer 3: Right, that’s pretty much what I’m wondering. We probably still want the cert-manager API, as used in ACM, but I don’t think we need to add our own syncing mechanism. I know the addons for example have CSRs, signed using cert-manager without any other syncing mechanism, but I haven’t looked into the details — in particular, whether spoke-to-hub CSRs are supported.

When the CA_CSR_Syncer picks up the Submariner CSR resource, it will issue a CA CSR towards the CA.
Monitor the result of the CA CSR and publish the result back to the original Gateway via the Broker
(by updating the original CSR CRD). The details of the design are shown in the upcoming sections.

## Design Details

### CA Deployment

The following diagram shows the various component needed for the central CA.

> ![CA Deployment](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/maryamtahhan/enhancements/feat_central_cert_manager/submariner/images/CA-deploy.puml)

Reviewer: After this is merged, we should remember to update these source URLs to reference the UMLs that will be merged as a part of this PR.

### CA Issue Certificate Example

> **_NOTE:_**
>
> * For Simplicity purposes the role of Admiral in the diagrams below has been
> collapsed into the CSR_Syncers.

![CA Issue Certificate Example](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/maryamtahhan/enhancements/feat_central_cert_manager/submariner/images/CA-operation.puml)

### Submariner CSR CRD

```Go
type SubmarinerCSRs struct {
    metav1.TypeMeta `json:",inline"`
    metav1.ObjectMeta `json:"metadata,omitempty"`
    
    // CSR towards the Submariner CA.
    CSRSpec SubmarinerCSRSpec `json:"spec"`

    // Result of CSR.
    CSRStatusSpec SubmarinerCSRStatus `json:"spec"`
}

// SubmarinerCSRSpec defines the CSR for the Submariner Gateway
type SubmarinerCSRSpec struct {
    // The requested 'duration' (i.e. lifetime) of the Certificate.
    Duration *metav1.Duration

>     // The PEM-encoded x509 certificate signing request to be submitted to the

Reviewer: ```suggestion
    // The PEM-encoded X.509 certificate signing request to be submitted to the
```

    // CA for signing.
    Request []byte

>     // Usages is the set of x509 usages that are requested for the certificate.

Reviewer: ```suggestion
    // Usages is the set of X.509 usages that are requested for the certificate.
```

    // "signing", "digital signature", "key agreement"...
    // details can be found here: https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.3
    // and https://github.com/cert-manager/cert-manager/blob/master/internal/apis/certmanager/types.go#L159

>     Usages [] string

Author: ```suggestion
    Usages []string
```

}

// SubmarinerCSRStatus defines the status of CertificateRequest and
// resulting signed certificate.
type SubmarinerCSRStatus struct {
    // Status of the request (`Ready`, `InvalidRequest`, `Approved`, `Denied`).
    Status string

>     // The PEM encoded x509 certificate resulting from the certificate

Reviewer: ```suggestion
    // The PEM-encoded X.509 certificate resulting from the certificate
```

    // signing request. Empty if the Status is InvalidRequest/Denied.
    Certificate []byte

>     // The PEM encoded x509 certificate of the Certificate Authority.

Reviewer: ```suggestion
    // The PEM-encoded X.509 certificate of the Certificate Authority.
```

    CA []byte
}
```

### CA CSR CRD

> The Proposed CA project to leverage is [cert-manager](https://cert-manager.io/docs/) Here is an

Reviewer: ```suggestion
The Proposed CA project to leverage is [cert-manager](https://cert-manager.io/docs/). Here is an
```
